test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,250 @@
+//---------------------------------------------------------------------
+// <copyright file="PreserveTypesAndAttributesTests.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.OData.Client.Tests.ALinq
+{
+    /// <summary>
+    /// Tests to check original resource type is used when retrieving attributes like OriginalNameAttribute
+    /// even if the IQueryable<T> object is casted to another type.
+    /// </summary>
+    public class PreserveTypesAndAttributesTests
+    {
+        private readonly TestContext ctx;
+
+        public PreserveTypesAndAttributesTests()
+        {
+            string uri = "https://services.odata.org/V4/Northwind/Northwind.svc";
+            this.ctx = new TestContext(new Uri(uri));
+        }
+
+        [Fact]
+        public void OneLevelCase1()
+        {
+            IQueryable<ICustomer> q1 = ctx.OneLevelCustomers1;
+            var r1 = q1.Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(5).ToList();
```

Comment:
Could you better name these variables so that its clear what they represent.

Reply:
Done

Reply:
@wachugamaina are the new names ok for you?